user-org-keycloak-util/src/main/java/org/sunbird/sso/service/impl/KeyCloakRsaKeyFetcher.java

```diff
@@ -0,0 +1,133 @@
+      if (publicKeyString != null) {
+        valueMap = getValuesFromJson(publicKeyString);
+        if (valueMap != null) {
+          BigInteger modulus = new BigInteger(1, urlDecoder.decode(valueMap.get(MODULUS)));
+          BigInteger publicExponent = new BigInteger(1, urlDecoder.decode(valueMap.get(EXPONENT)));
+          PublicKey key = keyFactory.generatePublic(new RSAPublicKeySpec(modulus, publicExponent));
+          saveToCache(key);
+          return key;
+        }
+      }
+    } catch (Exception e) {
+      ProjectLogger.log(
+          "KeyCloakRsaKeyFetcher:getPublicKeyFromKeyCloak: Exception occurred with message = "
+              + e.getMessage(),
+          LoggerEnum.ERROR);
+    }
+    return null;
+  }
+
+  /**
+   * This method will save the public key string value to cache
+   *
+   * @param key Public key to save in cache
+   */
+  private void saveToCache(PublicKey key) {
+    byte[] encodedPublicKey = key.getEncoded();
+    String publicKey = Base64.getEncoder().encodeToString(encodedPublicKey);
+    PropertiesCache cache = PropertiesCache.getInstance();
+    cache.saveConfigProperty(SSOConstant.SSO_PUBLIC_KEY, publicKey);
+  }
+
+  /**
+   * This method will connect to keycloak server using API call for getting public key.
+   *
+   * @param url A string value having keycloak base URL
+   * @param realm Keycloak realm name
+   * @return Public key JSON response string
+   */
+  private String requestKeyFromKeycloak(String url, String realm) {
+    HttpClient client = HttpClientBuilder.create().build();
+    HttpGet request = new HttpGet(url + "realms/" + realm + "/protocol/openid-connect/certs");
+
+    try {
+      HttpResponse response = client.execute(request);
+      HttpEntity entity = response.getEntity();
+
+      if (entity != null) {
+        return EntityUtils.toString(entity);
+      } else {
+        ProjectLogger.log(
+            "KeyCloakRsaKeyFetcher:requestKeyFromKeycloak: Not able to fetch SSO public key from keycloak server",
+            LoggerEnum.ERROR);
+      }
+    } catch (IOException e) {
+      ProjectLogger.log(
+          "KeyCloakRsaKeyFetcher:requestKeyFromKeycloak: Exception occurred with message = "
+              + e.getMessage(),
+          LoggerEnum.ERROR);
+    }
+    return null;
+  }
+
+  /**
+   * This method will return a map containing values extracted from public key JSON string.
+   *
+   * @param response Public key JSON response string
+   */
+  private Map<String, String> getValuesFromJson(String response) {
+    ObjectMapper mapper = new ObjectMapper();
+    Map<String, String> values = new HashMap<>();
+    try {
+      JsonNode res = mapper.readTree(response);
+      JsonNode keys = res.get("keys");
+      if (keys != null) {
+
+        JsonNode value = keys.get(0);
+        values.put(MODULUS, value.get("n").asText());
+        values.put(EXPONENT, value.get("e").asText());
+      }
+    } catch (Exception e) {
```

Comment:
instead of returning null , its good practice to return empty map i.e Collections.emptyMap()

Reply:
Here we are returning PublicKey object.